src/software/thunderscope/robot_diagnostics/robot_error_log.py

```diff
@@ -0,0 +1,196 @@
+        Doesn't log if this robot has already logged a specific error code type (to prevent spam)
+        Clears the types of error codes already added for a robot if no error
+        :param robot_status: the RobotStatus message
+        """
+        # initialize the list if not in dict yet
+        if robot_status.robot_id not in self.error_code_log_disabled:
+            self.error_code_log_disabled[robot_status.robot_id] = []
+
+        # if there's an error code, add to log
+        for code in robot_status.error_code:
+            if (
+                code != ErrorCode.NO_ERROR
+                and code != ErrorCode.LOW_BATTERY
+                and code in ERROR_CODE_MESSAGES.keys()
+                and code not in self.error_code_log_disabled[robot_status.robot_id]
+            ):
+                self.add_error_log_message(
+                    ErrorCodeLogMessageWidget(
+                        robot_status.robot_id, ERROR_CODE_MESSAGES[code]
+                    )
+                )
+                # add to list for this robot id to prevent spamming
+                self.error_code_log_disabled[robot_status.robot_id].append(code)
+            # if there is no more errors, clear the list so far
+            elif code == ErrorCode.NO_ERROR:
+                self.error_code_log_disabled[robot_status.robot_id] = []
+
+    def __refresh_robot_status_battery_voltage(self, robot_status: RobotStatus) -> None:
+        """
+        If the current robot's battery is lower than the threshold, adds a log
+        Records that this robot has already had a log added to prevent spamming
+        Resets the spamming flag once voltage for this robot goes back above threshold
+        :param robot_status: the RobotStatus message
+        """
+        # if the battery voltage is too low and a voltage log hasn't been sent for this robot yet, log it
+        if robot_status.power_status.battery_voltage <= BATTERY_WARNING_VOLTAGE and (
+            robot_status.robot_id not in self.low_battery_log_disabled
+            or not self.low_battery_log_disabled[robot_status.robot_id]
+        ):
+            self.add_error_log_message(
+                LowBatteryLogMessageWidget(robot_status.robot_id)
+            )
+            # prevent spamming logs by disabling logs
+            self.low_battery_log_disabled[robot_status.robot_id] = True
+        elif robot_status.power_status.battery_voltage > BATTERY_WARNING_VOLTAGE:
+            # battery voltage is now above threshold again, so start logging low battery
+            self.low_battery_log_disabled[robot_status.robot_id] = False
+
+    def add_error_log_message(self, error_widget: RobotLogMessageWidget) -> None:
+        """
+        Adds the given error message to the log by making a new RobotErrorMessage
+        :param error_widget: the error widget to add to the log
+        """
+        self.error_log_messages.append(error_widget)
```

Comment:
Is there a performance impact if we keep adding more error logs over time?

Does pyqt handle clearing memory associated with widgets that aren't being displayed anymore or does the user have to do it?

Reply:
right now they're not cleared, i can add a user close button and also clear them after a certain time